### ISSUE TITLE
HttpClientFactory holds onto exceptions indefinitely, despite a two-minute timer

### ISSUE THREAD

**Issue by amittleider** [dotnet/runtime#80605]
### Description

# Summary
I have opened a PR here: https://github.com/dotnet/runtime/pull/80604 to demonstrate this issue. The PR is not meant to be merged, it's merely a demonstration of the issue that I'm reporting. The test code runs and demonstrates the issue well. The production code is incomplete, but shows a potential path to fixing the issue.

The problem is that DefaultHttpClientFactory doesn't handle the case where an factory throws an exception well.

Typically, an HttpClientHandler will be recycled after two minutes (https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs#L499), but, in the case where the building of the handler throws an exception, the exception will be cached indefinitely. 

In the definition of a Lazy object:
```
Exception caching When you use factory methods, exceptions are cached. That is, if the factory method throws an exception the first time a thread tries to access the [Value](https://learn.microsoft.com/en-us/dotnet/api/system.lazy-1.value?view=net-7.0) property of the [Lazy<T>](https://learn.microsoft.com/en-us/dotnet/api/system.lazy-1?view=net-7.0) object, the same exception is thrown on every subsequent attempt.
```

This Lazy object is accessed in the `CreateHandler` method https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs#L118 , and immediately after it is accessed, the code will start a timer with the `StartHandlerEntry` timer. So, if line 118 throws an exception, a timer will never be set, and there will never be an attempt to re-initialize the object and the handler will be indefinitely in a bad state and irrecoverable.

# Example scenario
As an example of how this might occur, consider a factory which takes runtime configuration as input. The runtime configuration changes to a bad state, causing an exception in the handler. The runtime configuration is then updated to a good state, but the factory will never be called again, so the application must be restarted to clear the Lazy object from memory.

# Proposed solutions
1. One solution is to ensure that the HandlerTimer is set before the factory is called. This will ensure that even if exceptions occur, we will always try to re-initialize it after the two-minute default. In the current implementation, the timer depends on the ActiveHandlerTrackingEntry, so this would require a bit of a re-write in the way the timers are handled. (See the code in the PR https://github.com/dotnet/runtime/compare/main...amittleider:runtime:HttpClientFactory_ExceptionsPersist?expand=1#diff-940dba6ffc5ae548fca7105241869c1a78442c04b968f129ce645407022e83cbR118 . This code doesn't compile, but it helps to illustrate the proposal).
2. Another solution would be to make a new type of Lazy object that has a timer pre-built in. Like this, there is no need to handle timers at all within the DefaultHttpClientFactory.

### Reproduction Steps

See the test code in the PR: https://github.com/dotnet/runtime/pull/80604/files#diff-7ee446a98cb0ad2039642e909ac0732b37e7f764b534651cd88a3ac910c5b382R20

### Expected behavior

See the test code in the PR: https://github.com/dotnet/runtime/pull/80604/files#diff-7ee446a98cb0ad2039642e909ac0732b37e7f764b534651cd88a3ac910c5b382R45

### Actual behavior

See the test code in the PR: https://github.com/dotnet/runtime/pull/80604/files#diff-7ee446a98cb0ad2039642e909ac0732b37e7f764b534651cd88a3ac910c5b382R53

### Known Workarounds

Never throw exceptions in HttpClientHandlerFactories.

**Comment by amittleider** [comment]
> I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

@dotnet/area-extensions-dependencyinjection
@dotnet/ncl

**Comment by CarnaViire** [comment]
In case of `LazyThreadSafetyMode.PublicationOnly`, the initialization method can be executed concurrently multiple times per name, which is not acceptable perf-wise, as it would lead to unnecessary allocations (several message handlers being created instead of one). Handler configuration code, which is called within the initialization, can also rely on it being executed in a thread-safe manner, which would break.

If we were to modify the behavior here, we would probably need to let go of Lazy and implement synchronization manually by some other means.

**Comment by sbakharev** [comment]
Hello everyone.

> But it seems to be the first time we had this reported

We've just faced this issue, so I'll add my two cents :)

Our front-end service must use a client certificate for requests to back-end services. We load and set these in a configuration handler passed to `ConfigurePrimaryHttpMessageHandler` method. Certificates are taken from the local certificate store.

Since client certificates in our scenario are mandatory, the class which loads them from the store throws exceptions when things go wrong. We can't swallow these exceptions, because if we do that, all subsequent requests would result in 403 responses from back-ends, basically making the whole service non-operational.

In our particular case the problem emerged during client certificates rotation:

- A new certificate was installed into the system next to the old one
- The service started catching up the new certificate (it selects the one with bigger "not after" date/time when multiple certificates are found)
- An exception was thrown because by the time the new certificate has been found and selected, we still haven't granted the service user rights to read the certificate's private key
- Even though we did grant those rights about 1 minute after cert installation, the service was already dead and only restarting it could help (which we unfortunately didn't know at the moment, so lost more time)

Yes, we could make our certs loading/selecting strategy smarter to mitigate this particular problem (no private key rights _yet_), but that's just a workaround. And we would need more workarounds for other scenarios - certificate has disappeared, has expired, has been revoked, etc.

The issue here is with "design decision to expect configuration callbacks to be no-throw". Sometimes exceptions **must** be thrown from those callbacks when they operate in dynamic environments, otherwise you end up with a broken service. Certificate store is one example of such a dynamic environment - it is subject to regular changes in runtime - and to me it doesn't look like a corner-case scenario. IMO having a way to automatically recover here is not a *nice-to-have*, it is a *must-have* feature. Maybe we could think of an adequate way to inform the factory that the exception is potentially recoverable? Or maybe provide a retry policy?

The best option we have by far is to swallow all the exceptions and deal with a 2-minute downtime. Some may say it's not that scary, but... You know, 2 minutes may be quite valuable :)

P.S. 3 experienced developers barely could believe that this is how the framework actually behaves. A great indicator that such behavior is counter-intuitive.

**Comment by CarnaViire** [comment]
Thanks for the input @sbakharev!
This also made me thinking that the approach described in https://github.com/dotnet/runtime/issues/42143#issuecomment-3361602121 might help here too, on top of just allowing for faster/explicit handler cleanup.

**Comment by amittleider** [comment]
> Our front-end service must use a client certificate for requests to back-end services. We load and set these in a configuration handler passed to `ConfigurePrimaryHttpMessageHandler` method. Certificates are taken from the local certificate store.
> 
> Since client certificates in our scenario are mandatory, the class which loads them from the store throws exceptions when things go wrong. We can't swallow these exceptions, because if we do that, all subsequent requests would result in 403 responses from back-ends, basically making the whole service non-operational.
> 
> In our particular case the problem emerged during client certificates rotation:

This was also my scenario and reason for reporting the problem originally. Cert rotation is a common scenario and it will likely be implemented like this.

**Comment by CarnaViire** [comment]
FWIW cert rotation is usually better to implement using [`LocalCertificateSelectionCallback`](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslclientauthenticationoptions.localcertificateselectioncallback?view=net-9.0) that can refresh the certificate on each connection attempt, instead of using a single cert per the whole client/handler lifetime... e.g. see an updated example from https://github.com/dotnet/runtime/issues/52779:

```c#
services.AddHttpClient(name, ...)
    .UseSocketsHttpHandler((handler, provider) =>
    {​​​​​
        var certRepo = provider.GetRequiredService<ICertificateRepository>();

        // Using LocalCertificateSelectionCallback instead of the ClientCertificates collection
        // allows us to dynamically retrieve the certificate from the repository in case it has changed
        // NOTE: this callback executes on each connection creation
        handler.SslOptions.LocalCertificateSelectionCallback = delegate { return certRepo.GetCertificate(name); };
    });
```

How it solves the scenario:

- The handler is created successfully (without checking the cert)
- Cert is retrieved on the first request (= first connection attempt)
- Connection is aborted by the server if the cert is invalid
- In parallel, say, you update the database with a new, up to date cert
- The next request tries to establish a connection again, the handler picks up the new cert and proceeds successfully - so, neither a restart, nor even a new handler (= handler expiry = 2 min timer) is needed
